### PR TITLE
Fixes #1131 - Add E2E tests for checking the toggle button 

### LINKED ISSUE #1131: Add E2E test for toggling an indeterminate checkbox to true, false, true
## Context

See #1122 for more info and instructions on writing E2E tests.

## This test

Write an E2E test which asserts that "toggling an indeterminate checkbox to true, false, true" works as-expected. Here's a visual demo of what the test should do:

Peek 2022-03-02 17-09

### PR DESCRIPTION
**Pull request by Cronus1007**
Fixes #1131 

## Technical details

Adding E2E tests for checking the toggle button from intermediate to true,false and true.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by Cronus1007** [comment]
@seancolsen I am unable to identify any specific identifiers for the checkbox.Can you please help me out in it?

File: mathesar/tests/integration/test_toggle_button.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    expect(get_table_entry(page, "Table 0")).to_be_visible()
+    page.click("button:has-text('New Record')")
+    page.click("button:has-text('Refresh')")    
+    cell1 = page.goto(".s-sG5YPZwKUQz_ div :nth-child(5) input[type=checkbox]")
+    page.check(cell1)
+    assert page.is_checked(cell1) is True
+    page.uncheck(cell1)
+    assert page.is_checked(cell1) is False
+    page.check(cell1)
+    assert page.is_checked(cell1) is True
```
Comment (Cronus1007) [inline mathesar/tests/integration/test_toggle_button.py:25]: @seancolsen @pavish @kgodey Shall I add a new line at the end of file?

Comment (silentninja) [reply mathesar/tests/integration/test_toggle_button.py:25]: Newline at the end of file is needed to pass flake8 W292 rule, so yes it should be added

**Comment by Cronus1007** [comment]
@silentninja @seancolsen @pavish When I run these tests in playwright inspector they works fine but fails in ci/cd pipelines any particular reason for this.

**Comment by silentninja** [comment]
@Cronus1007 The [last workflow](https://github.com/centerofci/mathesar/runs/5641602430?check_suite_focus=true) is failing due to linting issues. Fixing those should help in passing CI pipeline

**Comment by Cronus1007** [comment]
@silentninja I have fixed this workflow error in the commit `f41a3dc`. I guess ci/cd pipelines should work now.

File: mathesar/tests/integration/test_toggle_button.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    expect(get_table_entry(page, "Table 0")).to_be_visible()
+    page.click("button:has-text('New Record')")
+    page.click("button:has-text('Refresh')")
+    cell1 = page.locator("div:nth-child(5) div:nth-child(3) .cell-wrapper input[type=checkbox]")
+    cell1.check()
+    assert page.is_checked("div:nth-child(5) div:nth-child(3) .cell-wrapper input[type=checkbox]") is True
+    cell1.uncheck()
+    assert page.is_checked("div:nth-child(5) div:nth-child(3) .cell-wrapper input[type=checkbox]") is False
+    cell1.check()
+    assert page.is_checked("div:nth-child(5) div:nth-child(3) .cell-wrapper input[type=checkbox]") is True
```
Comment (Cronus1007) [inline mathesar/tests/integration/test_toggle_button.py:24]: @seancolsen @silentninja When I run these tests locally all the tests passes but when I try to run it on ci/cd pipelines it fails. Can you please help me out of it?

**Comment by seancolsen** [comment]
@Cronus1007 

> When I run these tests locally all the tests passes but when I try to run it on ci/cd pipelines it fails

I'm noticing something similar on my end. When running Mathesar via the local development environment, I'm able to create a table by copy-pasting the text used for this E2E test. But when I run the E2E test locally in headed mode, I notice that the column type is getting set to TEXT instead of BOOLEAN. Within the E2E browser, I can try to manually set the type to BOOLEAN, and I get this error:

```
(psycopg2.errors.InvalidSchemaName) schema "mathesar_types" does not exist
LINE 1: SELECT table_tests."Table 0".id, mathesar_types.cast_to_bool...
                                         ^

[SQL: SELECT table_tests."Table 0".id, mathesar_types.cast_to_boolean(table_tests."Table 0".foo) AS foo 
FROM table_tests."Table 0" 
 LIMIT %(param_1)s]
[parameters: {'param_1': 20}]
(Background on this error at: http://sqlalche.me/e/14/f405)
```

That leads me to believe that #1168 is causing this problem. I think we should hold off on continuing work on this PR until that ticket is closed.

**Comment by Cronus1007** [comment]
@seancolsen Sure until then I will work upon other issues that have been assigned to me.

**Comment by seancolsen** [comment]
> I will work upon other issues

Sounds good.

> Shall I also close this pr?

No, please keep this PR open. We'll come back to it once #1168 is finished.

**Comment by pavish** [comment]
@seancolsen @Cronus1007 

Is there a reason why we are creating a new table in this test through UI interactions, as it is unrelated to the test itself?

We can use an existing table with boolean columns (eg., add a boolean column to the `table_with_all_types` and use it here). This should make this PR unblocked.

**Comment by seancolsen** [comment]
> Is there a reason why we are creating a new table in this test through UI interactions, as it is unrelated to the test itself?

Yeah good question. That's just how I spec'ed out the ticket, because we didn't yet have `table_with_all_types`, and I wanted to keep the specs simple for new contributors.

> add a boolean column to the `table_with_all_types` and use it here

I think that's a good idea.

@Cronus1007 You can proceed in that direction if you like. You can see the tests in `test_text_type.py` as examples of tests to model yours after. And in `mathesar/tests/integration/conftest.py`, you'll need modify the `table_with_all_types` function to add a boolean column and add a row to the table with a null value in the boolean column.

**Comment by Cronus1007** [comment]
@pavish @seancolsen Sure I would like to go ahead in this manner.

**Comment by Cronus1007** [comment]
> @Cronus1007 You can proceed in that direction if you like. You can see the tests in `test_text_type.py` as examples of tests to model yours after. And in `mathesar/tests/integration/conftest.py`, you'll need modify the `table_with_all_types` function to add a boolean column and add a row to the table with a null value in the boolean column.

@seancolsen Shall I drop this idea now since #1168 is already finished and proceed in the manner the ticket was supposed to be.

File: mathesar/tests/integration/test_toggle_button.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+from mathesar.tests.integration.utils.locators import get_table_entry, get_tables_list
+
+
+def test_toggle_button(page, base_schema_url):
+    page.goto(base_schema_url)
+    expect(get_tables_list(page)).to_be_empty()
+    page.click("[aria-label='New Table']")
+    page.click("button:has-text('Import Data')")
+    page.click("text=Copy and Paste Text")
+    page.fill("textarea", "foo\ntrue\nfalse\nfalse\ntrue")
```
Comment (Cronus1007) [inline mathesar/tests/integration/test_toggle_button.py:12]: @seancolsen While running this line on playwright inspector playwright treats it as a text and hence the type becomes TEXT while running it locally while I do the same thing locally the browser treats this as a BOOLEAN type. Can you plzzz tell me how to tackle this situation 

Playwright Inspector
Screenshot from 2022-03-29 14-55-26
Browser
Screenshot from 2022-03-29 14-55-34

**Comment by seancolsen** [comment]
@Cronus1007 

> @seancolsen Shall I drop this idea now since https://github.com/centerofci/mathesar/issues/1168 is already finished and proceed in the manner the ticket was supposed to be.

I'm fine with either approach -- the approach one described by the screencap in the original ticket, or the approach described by Pavish's [comment](https://github.com/centerofci/mathesar/pull/1204#issuecomment-1076787436). If you don't care, then I'd recommend Pavish's approach because it's cleaner.

File: mathesar/tests/integration/test_toggle_button.py (hunk 3 above)
Comment (seancolsen) [reply mathesar/tests/integration/test_toggle_button.py:12]: My understanding is that #1168 is supposed to fix that. Are you running with the latest code?

Comment (Cronus1007) [reply mathesar/tests/integration/test_toggle_button.py:12]: @seancolsen Yes I am running it with the latest code.

**Comment by Cronus1007** [comment]
@seancolsen I have implemented @pavish approach. Once I test them locally I would update the PR.

**Comment by Cronus1007** [comment]
@seancolsen @pavish @silentninja This PR is now ready for the review. @seancolsen Shall I also record how the tests work on playwright inspector and attach it here or is it not required that much.

**Review by seancolsen — APPROVED** [review]
Thanks @Cronus1007 

I added another commit with some improvements. You can see my commit message for details.

**Pull request merged**